### PR TITLE
DO_NOT_MERGE: [FEATURE] General Robust Serialization Utility for Configuration Objects that May Contain DataFrames

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
In order to deal with non-serializable values of various attributes inside configuration objects, we introduce a general serialization utility to be utilized in such situations.

The mechanism for serializing configuration objects that contain non-serializable values, such as Pandas and Spark dataframes consists of the following steps:
1) Scan the original configuration object and mark the specified attributes (that reference non-serializable values)
with temporary unique hexadecimal value strings, while maintaining the reference map between these temporary labels and
the original (i.e., attributed as non-serializable) values.
2) The updated configuration object should now be entirely serializable, because all the attributed non-serializable
values have been temporarily replaced by their corresponding unique string labels.
3) Restore the original configuration object and any copies thereof using the reference map built in the first step.

Example:
```
        reference_map: Dict[str, Dict[str, Any]] = {}
        mark_and_replace_non_serializable_references_in_config(
            config=obj,
            attribute_names={"batch_data"},
            reference_map=reference_map,
        )
        data: dict = super().dump(obj=obj, many=many)  #  or: data: dict = copy.deepcopy(self.to_dict()) -- etc.
        restore_non_serializable_references_into_config(
            config=obj,
            attribute_names={"batch_data"},
            reference_map=reference_map,
        )
        restore_non_serializable_references_into_config(
            config=data,
            attribute_names={"batch_data"},
            reference_map=reference_map,
        )
```
and as part of a `__repr()__` type method also add:

```
            restore_non_serializable_references_into_config(
                config=data,
                attribute_names={"batch_data"},
                reference_map=reference_map,
                replace_value_with_type_string=True,
            )
            return json.dumps(data, indent=2)
```

Tests have been updated.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: ANT-20/ANT-22/ANT-66
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/core/batch.py [hunk 1]
```diff
@@ -822,157 +835,154 @@ def get_batch_request_dict(
...
 
-def get_runtime_parameters_batch_data_references_from_config(
-    config: Union[DictDot, dict]
-) -> Tuple[Optional[Any], Optional[List[Any]]]:
-    if not isinstance(config, (DictDot, dict)):
-        raise TypeError(
-            f"""The Checkpoint configuraiton argument must have the type "dict" or "DictDot" (the type given \
-is "{str(type(config))}", which is illegal).
 """
+The mechanism for serializing configuration objects that contain non-serializable values, such as Pandas and Spark
```
Comment (Shinnnyshinshin) [inline great_expectations/core/batch.py:839]: Thank you very very much for this docstring

File: great_expectations/core/batch.py [hunk 2]
```diff
@@ -822,157 +835,154 @@ def get_batch_request_dict(
...
+        for key in attribute_names:
+            reference_map[key] = {}
+
+    serializable_value: str
+
+    if isinstance(config, (DictDot, dict)):
+        for key, value in config.items():
+            if key in attribute_names:
+                serializable_value = (
+                    f'{key}_{hashlib.md5(key.encode("utf-8")).hexdigest()}'
```
Comment (Shinnnyshinshin) [inline great_expectations/core/batch.py:916]: :heart:

**Review by Shinnnyshinshin — APPROVED** [review]
Thank you for this @alexsherstinsky I think it's a really elegant solution to a really hairy problem... And I also appreciate the example and description in the PR. It helped me get oriented in the code a lot quicker 🙇🏼

**Comment by alexsherstinsky** [comment]
There are still serialization errors, which must be investigated and fixed.

**Pull request closed without merging**